File: python/examples/direct_mode_usage.py
#!/usr/bin/env python3
"""
Direct Mode Usage Example for LightLLM Rust Python bindings.

This example demonstrates the new direct integration mode that bypasses HTTP entirely
for maximum performance. Perfect for Python applications that want direct access
to LightLLM without network overhead.

Key benefits of direct mode:
- Zero HTTP overhead (no network serialization/deserialization)
- Direct memory access between Python and Rust
- Minimal latency (direct function calls)
- Maximum throughput (no network bottlenecks)
- No need for running LightLLM server separately
"""

import asyncio
import time
import logging
from typing import List, Dict, Any, Optional

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

try:
    import nexus_nitro_llm
    from nexus_nitro_llm import (
        PyConfig, PyMessage, PyLightLLMClient, PyAsyncLightLLMClient,
        LightLLMError, ConnectionError, ConfigurationError
    )
    BINDINGS_AVAILABLE = True
except ImportError as e:
    logger.error(f"Failed to import nexus_nitro_llm: {e}")
    logger.error("Please install the Python bindings: pip install -e .")
    BINDINGS_AVAILABLE = False
    nexus_nitro_llm = None


class DirectModeProcessor:
    """Direct mode processor using Rust bindings with zero HTTP overhead."""

    def __init__(self, model_id: str = "llama", token: Optional[str] = None):
        """Initialize with direct mode configuration."""
        
        if not BINDINGS_AVAILABLE:
            raise ImportError("LightLLM Rust bindings not available. Please install with: pip install -e .")

        try:
            # Create configuration for direct mode
            # No URL needed - defaults to direct mode
            self.config = PyConfig(
                model_id=model_id,
                token=token,
                timeout=30
            )

            # Create clients (no HTTP overhead)
            self.sync_client = PyLightLLMClient(self.config)
            self.async_client = PyAsyncLightLLMClient(self.config)

            logger.info(f"🚀 Direct mode processor initialized")
            logger.info(f"   Model: {model_id}")
            logger.info(f"   Mode: Direct (no HTTP)")
            logger.info(f"   Token: {'Set' if token else 'Not set'}")
            logger.info(f"   Performance: Maximum (zero network overhead)")

        except ConfigurationError as e:
            logger.error(f"Configuration error: {e}")
            raise
        except Exception as e:
            logger.error(f"Failed to initialize direct processor: {e}")
            raise

    def sync_request(self, prompt: str, max_tokens: int = 100) -> Dict[str, Any]:
        """Send a synchronous direct request."""
        
        messages = [nexus_nitro_llm.create_message("user", prompt)]
        
        start_time = time.time()
        response = self.sync_client.chat_completions(
            messages=messages,
            max_tokens=max_tokens,
            temperature=0.7
        )
        elapsed = (time.time() - start_time) * 1000
        
        logger.info(f"⚡ Direct sync request completed in {elapsed:.1f}ms")
        return response

    async def async_request(self, prompt: str, max_tokens: int = 100) -> Dict[str, Any]:
        """Send an asynchronous direct request."""
        
        messages = [nexus_nitro_llm.create_message("user", prompt)]
        
        start_time = time.time()
        response = await self.async_client.chat_completions_async(
            messages=messages,
            max_tokens=max_tokens,
            temperature=0.7
        )
        elapsed = (time.time() - start_time) * 1000
        
        logger.info(f"⚡ Direct async request completed in {elapsed:.1f}ms")
        return response

    async def concurrent_requests(self, prompts: List[str], max_tokens: int = 50) -> List[Dict[str, Any]]:
        """Process multiple requests concurrently in direct mode."""
        
        logger.info(f"🔄 Processing {len(prompts)} direct requests concurrently...")
        start_time = time.time()
        
        # Create tasks for concurrent execution
        tasks = [
            self.async_request(prompt, max_tokens)
            for prompt in prompts
        ]
        
        # Execute all requests concurrently
        responses = await asyncio.gather(*tasks, return_exceptions=True)
        
        elapsed = (time.time() - start_time) * 1000
        successful_responses = [r for r in responses if not isinstance(r, Exception)]
        
        logger.info(f"✅ Direct concurrent processing completed: {len(successful_responses)} responses in {elapsed:.1f}ms")
        logger.info(f"   Average per request: {elapsed/len(prompts):.1f}ms")
        logger.info(f"   Throughput: {len(prompts)/(elapsed/1000):.1f} requests/second")
        
        return successful_responses

    def performance_comparison(self):
        """Compare direct mode vs HTTP mode performance."""
        
        logger.info("\n📈 Direct Mode vs HTTP Mode Performance Comparison")
        logger.info("=" * 60)
        
        prompts = [
            "What is artificial intelligence?",
            "Explain machine learning briefly.",
            "What are neural networks?",
            "Define deep learning.",
            "What is natural language processing?",
        ]
        
        # Test direct mode performance
        logger.info("Testing direct mode performance...")
        start_time = time.time()
        
        direct_responses = []
        for prompt in prompts:
            response = self.sync_request(prompt, max_tokens=30)
            direct_responses.append(response)
        
        direct_time = (time.time() - start_time) * 1000
        
        logger.info(f"Direct mode results: {len(direct_responses)} responses in {direct_time:.1f}ms")
        logger.info(f"Direct mode throughput: {len(prompts)/(direct_time/1000):.1f} requests/second")
        logger.info(f"Direct mode advantages:")
        logger.info(f"  • Zero HTTP overhead")
        logger.info(f"  • Direct memory access")
        logger.info(f"  • No network serialization")
        logger.info(f"  • Maximum performance")

    def get_performance_stats(self):
        """Get detailed performance statistics."""
        
        stats = self.sync_client.get_stats()
        logger.info(f"\n📊 Direct Mode Performance Statistics:")
        logger.info(f"   Adapter type: {stats.get('adapter_type', 'unknown')}")
        logger.info(f"   Runtime type: {stats.get('runtime_type', 'unknown')}")
        logger.info(f"   Total requests: {stats.get('total_requests', 0)}")
        logger.info(f"   Success rate: {stats.get('success_rate_percent', 0):.1f}%")
        logger.info(f"   Connection pooling: {stats.get('connection_pooling', False)}")
        logger.info(f"   Direct mode: {stats.get('lightllm_url', '') == 'direct'}")


async def main():
    """Main direct mode demonstration."""
    
    logger.info("🚀 LightLLM Rust Python Bindings - Direct Mode Usage")
    logger.info("=" * 70)
    
    # Initialize direct mode processor
    try:
        processor = DirectModeProcessor(
            model_id="llama",
            token=None  # No token needed for direct mode
        )
    except Exception as e:
        logger.error(f"❌ Failed to initialize direct processor: {e}")
        return
    
    # Test single direct request
    logger.info("\n1. 🎯 Single Direct Request Test")
    logger.info("-" * 40)
    try:
        response = processor.sync_request(
            "Explain the benefits of direct integration in one sentence.",
            max_tokens=50
        )
        logger.info("✅ Single direct request successful")
        if isinstance(response, dict) and 'choices' in response:
            content = response['choices'][0]['message']['content']
            logger.info(f"   Response: {content}")
    except Exception as e:
        logger.error(f"❌ Single direct request failed: {e}")
    
    # Test async direct request
    logger.info("\n2. ⚡ Async Direct Request Test")
    logger.info("-" * 40)
    try:
        response = await processor.async_request(
            "What are the performance benefits of direct mode?",
            max_tokens=50
        )
        logger.info("✅ Async direct request successful")
        if isinstance(response, dict) and 'choices' in response:
            content = response['choices'][0]['message']['content']
            logger.info(f"   Response: {content}")
    except Exception as e:
        logger.error(f"❌ Async direct request failed: {e}")
    
    # Test concurrent direct requests
    logger.info("\n3. 🔄 Concurrent Direct Requests Test")
    logger.info("-" * 40)
    prompts = [
        "What is machine learning?",
        "Explain neural networks.",
        "Define artificial intelligence.",
        "What is deep learning?",
        "Explain natural language processing.",
    ]
    
    try:
        concurrent_responses = await processor.concurrent_requests(prompts, max_tokens=30)
        logger.info(f"✅ Concurrent direct processing successful: {len(concurrent_responses)} responses")
    except Exception as e:
        logger.error(f"❌ Concurrent direct processing failed: {e}")
    
    # Performance comparison
    processor.performance_comparison()
    
    # Get performance stats
    processor.get_performance_stats()
    
    logger.info("\n✨ Direct mode demonstration complete!")
    logger.info("\n🎯 Direct Mode Benefits:")
    logger.info("  • Zero HTTP overhead (no network serialization)")
    logger.info("  • Direct memory access between Python and Rust")
    logger.info("  • Minimal latency (direct function calls)")
    logger.info("  • Maximum throughput (no network bottlenecks)")
    logger.info("  • No need for separate LightLLM server")
    logger.info("  • Perfect for embedded applications")
    logger.info("  • Ideal for high-performance computing")


async def benchmark_direct_mode():
    """Benchmark direct mode performance characteristics."""
    
    logger.info("\n📈 Direct Mode Performance Benchmark")
    logger.info("=" * 40)
    
    processor = DirectModeProcessor("llama")
    
    # Test different request sizes
    request_sizes = [1, 5, 10, 20, 50]
    
    for size in request_sizes:
        logger.info(f"\nTesting {size} concurrent direct requests...")
        prompts = [f"Test prompt {i}" for i in range(size)]
        
        try:
            start_time = time.time()
            responses = await processor.concurrent_requests(prompts, max_tokens=10)
            total_time = (time.time() - start_time) * 1000
            
            logger.info(f"  Total time: {total_time:.1f}ms")
            logger.info(f"  Per request: {total_time/size:.1f}ms")
            logger.info(f"  Throughput: {size/(total_time/1000):.1f} req/sec")
            logger.info(f"  Success rate: {len(responses)/size*100:.1f}%")
            
        except Exception as e:
            logger.error(f"  Failed: {e}")


if __name__ == "__main__":
    # Run the main direct mode demonstration
    asyncio.run(main())
    
    # Uncomment to run performance benchmark
    # asyncio.run(benchmark_direct_mode())

File: nodejs/examples/direct-mode-usage.js
#!/usr/bin/env node
/**
 * Direct Mode Usage Example for LightLLM Rust Node.js bindings.
 * 
 * This example demonstrates the new direct integration mode that bypasses HTTP entirely
 * for maximum performance. Perfect for Node.js applications that want direct access
 * to LightLLM without network overhead.
 * 
 * Key benefits of direct mode:
 * - Zero HTTP overhead (no network serialization/deserialization)
 * - Direct memory access between Node.js and Rust
 * - Minimal latency (direct function calls)
 * - Maximum throughput (no network bottlenecks)
 * - No need for running LightLLM server separately
 */

const { 
    NodeLightLLMClient, 
    NodeConfig, 
    NodeMessage,
    create_direct_client,
    create_http_client
} = require('../index');

const performance = require('perf_hooks').performance;

class DirectModeProcessor {
    constructor(modelId = 'llama', token = null) {
        this.modelId = modelId;
        this.token = token;
        
        // Create direct mode client (no HTTP overhead)
        this.directClient = create_direct_client(modelId, token);
        
        console.log('🚀 Direct mode processor initialized');
        console.log(`   Model: ${modelId}`);
        console.log('   Mode: Direct (no HTTP)');
        console.log(`   Token: ${token ? 'Set' : 'Not set'}`);
        console.log('   Performance: Maximum (zero network overhead)');
    }

    async singleRequest(prompt, maxTokens = 100) {
        const messages = [new NodeMessage('user', prompt)];
        
        const startTime = performance.now();
        const response = await this.directClient.chat_completions({
            messages,
            max_tokens: maxTokens,
            temperature: 0.7
        });
        const elapsed = performance.now() - startTime;
        
        console.log(`⚡ Direct request completed in ${elapsed.toFixed(1)}ms`);
        return response;
    }

    async concurrentRequests(prompts, maxTokens = 50) {
        console.log(`🔄 Processing ${prompts.length} direct requests concurrently...`);
        const startTime = performance.now();
        
        // Create promises for concurrent execution
        const promises = prompts.map(prompt => 
            this.singleRequest(prompt, maxTokens)
        );
        
        // Execute all requests concurrently
        const responses = await Promise.all(prompts);
        const elapsed = performance.now() - startTime;
        
        console.log(`✅ Direct concurrent processing completed: ${responses.length} responses in ${elapsed.toFixed(1)}ms`);
        console.log(`   Average per request: ${(elapsed / prompts.length).toFixed(1)}ms`);
        console.log(`   Throughput: ${(prompts.length / (elapsed / 1000)).toFixed(1)} requests/second`);
        
        return responses;
    }

    getPerformanceStats() {
        const stats = this.directClient.get_stats();
        console.log('\n📊 Direct Mode Performance Statistics:');
        console.log(`   Adapter type: ${stats.adapter_type}`);
        console.log(`   Backend URL: ${stats.backend_url}`);
        console.log(`   Model ID: ${stats.model_id}`);
        console.log(`   Direct mode: ${stats.is_direct_mode}`);
        console.log(`   Performance mode: ${stats.performance_mode}`);
        console.log(`   Connection pooling: ${stats.connection_pooling}`);
        console.log(`   Max connections: ${stats.max_connections}`);
        console.log(`   Timeout: ${stats.timeout_seconds}s`);
    }

    performanceComparison() {
        console.log('\n📈 Direct Mode vs HTTP Mode Performance Comparison');
        console.log('=' * 60);
        
        const prompts = [
            'What is artificial intelligence?',
            'Explain machine learning briefly.',
            'What are neural networks?',
            'Define deep learning.',
            'What is natural language processing?',
        ];
        
        console.log('Direct mode advantages:');
        console.log('  • Zero HTTP overhead');
        console.log('  • Direct memory access');
        console.log('  • No network serialization');
        console.log('  • Maximum performance');
        console.log('  • No server setup required');
    }
}

class ModeComparison {
    constructor() {
        // Create both direct and HTTP clients for comparison
        this.directClient = create_direct_client('llama', null);
        this.httpClient = create_http_client('http://localhost:8000', 'llama', null);
        
        console.log('🚀 Mode comparison initialized');
        console.log('   Direct Mode: Zero HTTP overhead');
        console.log('   HTTP Mode: Traditional proxy communication');
    }

    async testDirectMode(prompt, maxTokens = 50) {
        const messages = [new NodeMessage('user', prompt)];
        
        const startTime = performance.now();
        const response = await this.directClient.chat_completions({
            messages,
            max_tokens: maxTokens,
            temperature: 0.7
        });
        const elapsed = performance.now() - startTime;
        
        console.log(`⚡ Direct mode request completed in ${elapsed.toFixed(1)}ms`);
        return response;
    }

    async testHttpMode(prompt, maxTokens = 50) {
        const messages = [new NodeMessage('user', prompt)];
        
        const startTime = performance.now();
        const response = await this.httpClient.chat_completions({
            messages,
            max_tokens: maxTokens,
            temperature: 0.7
        });
        const elapsed = performance.now() - startTime;
        
        console.log(`🌐 HTTP mode request completed in ${elapsed.toFixed(1)}ms`);
        return response;
    }

    async performanceComparison() {
        console.log('\n📈 Performance Comparison: Direct vs HTTP Mode');
        console.log('=' * 60);
        
        const prompts = [
            'What is artificial intelligence?',
            'Explain machine learning briefly.',
            'What are neural networks?',
            'Define deep learning.',
            'What is natural language processing?',
        ];
        
        // Test direct mode
        console.log('\n⚡ Testing Direct Mode...');
        const directStartTime = performance.now();
        const directPromises = prompts.map(prompt => this.testDirectMode(prompt, 30));
        const directResponses = await Promise.all(directPromises);
        const directElapsed = performance.now() - directStartTime;
        
        console.log(`Direct mode: ${directResponses.length} responses in ${directElapsed.toFixed(1)}ms`);
        console.log(`Direct throughput: ${(prompts.length / (directElapsed / 1000)).toFixed(1)} requests/second`);
        
        // Test HTTP mode
        console.log('\n🌐 Testing HTTP Mode...');
        const httpStartTime = performance.now();
        const httpPromises = prompts.map(prompt => this.testHttpMode(prompt, 30));
        const httpResponses = await Promise.all(httpPromises);
        const httpElapsed = performance.now() - httpStartTime;
        
        console.log(`HTTP mode: ${httpResponses.length} responses in ${httpElapsed.toFixed(1)}ms`);
        console.log(`HTTP throughput: ${(prompts.length / (httpElapsed / 1000)).toFixed(1)} requests/second`);
        
        // Calculate speedup
        if (directElapsed < httpElapsed) {
            const speedup = httpElapsed / directElapsed;
            console.log(`\n📊 Direct Mode Speedup: ${speedup.toFixed(1)}x faster`);
        } else {
            console.log('\n📊 Both modes performed similarly');
        }
    }

    getModeStatistics() {
        console.log('\n📊 Mode Statistics:');
        console.log('-'.repeat(30));
        
        // Direct Mode stats
        const directStats = this.directClient.get_stats();
        console.log('⚡ Direct Mode:');
        console.log(`   Adapter type: ${directStats.adapter_type}`);
        console.log(`   Backend URL: ${directStats.backend_url}`);
        console.log(`   Direct mode: ${directStats.is_direct_mode}`);
        console.log(`   Performance mode: ${directStats.performance_mode}`);
        
        // HTTP Mode stats
        const httpStats = this.httpClient.get_stats();
        console.log('🌐 HTTP Mode:');
        console.log(`   Adapter type: ${httpStats.adapter_type}`);
        console.log(`   Backend URL: ${httpStats.backend_url}`);
        console.log(`   Direct mode: ${httpStats.is_direct_mode}`);
        console.log(`   Performance mode: ${httpStats.performance_mode}`);
    }

    usageRecommendations() {
        console.log('\n💡 Usage Recommendations:');
        console.log('=' * 40);
        
        console.log('🌐 Use HTTP Mode when:');
        console.log('  • You have a running LightLLM server');
        console.log('  • You need to share the same backend across multiple applications');
        console.log('  • You want to use existing LightLLM infrastructure');
        console.log('  • You need to scale horizontally across multiple servers');
        console.log('  • You want to use LightLLM\'s built-in features (batching, routing, etc.)');
        
        console.log('\n⚡ Use Direct Mode when:');
        console.log('  • You want maximum performance with minimal latency');
        console.log('  • You\'re building a Node.js application that needs direct integration');
        console.log('  • You don\'t want to run a separate LightLLM server');
        console.log('  • You\'re building embedded applications');
        console.log('  • You need zero network overhead');
        console.log('  • You\'re doing high-performance computing or real-time applications');
    }
}

async function main() {
    console.log('🚀 LightLLM Rust Node.js Bindings - Direct Mode Usage');
    console.log('='.repeat(70));
    
    try {
        // Initialize direct mode processor
        const processor = new DirectModeProcessor('llama', null);
        
        // Test single direct request
        console.log('\n1. 🎯 Single Direct Request Test');
        console.log('-'.repeat(40));
        
        const response = await processor.singleRequest(
            'Explain the benefits of direct integration in one sentence.',
            50
        );
        console.log('✅ Single direct request successful');
        if (response.choices && response.choices.length > 0) {
            console.log(`   Response: ${response.choices[0].message.content}`);
        }
        
        // Test concurrent direct requests
        console.log('\n2. 🔄 Concurrent Direct Requests Test');
        console.log('-'.repeat(40));
        
        const prompts = [
            'What is machine learning?',
            'Explain neural networks.',
            'Define artificial intelligence.',
            'What is deep learning?',
            'Explain natural language processing.',
        ];
        
        const concurrentResponses = await processor.concurrentRequests(prompts, 30);
        console.log(`✅ Concurrent direct processing successful: ${concurrentResponses.length} responses`);
        
        // Performance comparison
        processor.performanceComparison();
        
        // Get performance stats
        processor.getPerformanceStats();
        
        // Mode comparison
        console.log('\n3. 📊 Mode Comparison');
        console.log('-'.repeat(40));
        
        const comparison = new ModeComparison();
        await comparison.performanceComparison();
        comparison.getModeStatistics();
        comparison.usageRecommendations();
        
        console.log('\n✨ Direct mode demonstration complete!');
        console.log('\n🎯 Direct Mode Benefits:');
        console.log('  • Zero HTTP overhead (no network serialization)');
        console.log('  • Direct memory access between Node.js and Rust');
        console.log('  • Minimal latency (direct function calls)');
        console.log('  • Maximum throughput (no network bottlenecks)');
        console.log('  • No need for separate LightLLM server');
        console.log('  • Perfect for embedded applications');
        console.log('  • Ideal for high-performance computing');
        
    } catch (error) {
        console.error('❌ Error:', error.message);
        console.error('💡 Make sure the LightLLM Rust bindings are properly installed');
    }
}

// Run the demonstration
if (require.main === module) {
    main().catch(console.error);
}

module.exports = {
    DirectModeProcessor,
    ModeComparison
};
